### PR TITLE
Implemented UPDATE

### PR DESCRIPTION
**Pull request by Hydrocharged**
Take a look at this. I have yet to implement all of the tests, as this is mainly to check the logic to make sure that I'm on the correct track and all. First time working with their pipeline stuff, it's kind of interesting.

### CODE REVIEW

File: engine_test.go [hunk 1]
```diff
@@ -2209,6 +2209,63 @@ func TestReplaceIntoErrors(t *testing.T) {
...
 }
 
+func TestUpdate(t *testing.T) {
+	var updates = []struct {
+		updateQuery    string
+		expectedUpdate []sql.Row
+		selectQuery     string
+		expectedSelect  []sql.Row
+	}{
+		{
```
Comment (zachmu) [inline engine_test.go]: Additional test cases you need:
* Setting the same column more than once. Copy MySQL's behavior for this
* Setting a non-existent column
* Setting null for a non-nullable column

File: engine_test.go [hunk 2]
```diff
@@ -2209,6 +2209,63 @@ func TestReplaceIntoErrors(t *testing.T) {
...
+			"UPDATE mytable SET s = 'updated row' WHERE i <> 9999;",
+			[]sql.Row{{int64(3), int64(3)}},
+			"SELECT * FROM mytable;",
+			[]sql.Row{{int64(1), "updated row"},{int64(2), "updated row"},{int64(3), "updated row"}},
+		},
+		{
+			"UPDATE floattable SET f32 = f32 + f32, f64 = f32 * f64 WHERE i = 2;",
+			[]sql.Row{{int64(1), int64(1)}},
+			"SELECT * FROM floattable WHERE i = 2;",
+			[]sql.Row{{int64(2), float32(3.0), float64(4.5)}},
```
Comment (zachmu) [inline engine_test.go]: This isn't the right result and indicates you have a bug. The second value should be 1.5 * 1.5. Looks like your column updates aren't independent.

I would write some additional tests for setting multiple columns in the same updates statement once you fix this.

File: sql/expression/set.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"fmt"
+	"github.com/src-d/go-mysql-server/sql"
+	"gopkg.in/src-d/go-errors.v1"
+)
+
+var errCannotSetField = errors.NewKind("Expected GetField expression on left and but %T")
+
+// Set updates the value of a field from a row.
+type Set struct {
```
Comment (zachmu) [inline sql/expression/set.go]: Would call this SetField (Set is too overloaded). Or SetColumn if Column is used over Field in other code.

File: engine_test.go (hunk 1 above)
Comment (zachmu) [reply engine_test.go]: Also:
* Updates involving limit and offset, combined with order by
* Tests without where clauses (can be combined with above)
* Tests where not all matched rows were updates (you have a test where no rows were updated, but not one where some rows were updated and some were not)

Comment (zachmu) [reply engine_test.go]: Also the verification queries aren't sufficient. Consider a hypothetical bug where the where clause was being ignored and every row updated. You need at least one tests that asserts the contents of an entire table after an update.

File: sql/plan/update.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+			if !equals {
+				err = updatable.Update(ctx, oldRow, newRow)
+				if err != nil {
+					_ = iter.Close()
+					return rowsMatched, rowsUpdated, err
+				}
+				rowsUpdated++
+			}
+		} else {
+			return rowsMatched, rowsUpdated, err
```
Comment (zachmu) [inline sql/plan/update.go]: Not closing the iterator in this case

File: sql/plan/update.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	pr := sql.NewTreePrinter()
+	_ = pr.WriteNode("Update")
+	_ = pr.WriteChildren(p.Node.String())
+	for _, updateExpr := range p.UpdateExprs {
+		_ = pr.WriteChildren(updateExpr.String())
+	}
+	return pr.String()
+}
+
+func (p *Update) applyUpdates(ctx *sql.Context, row sql.Row) (sql.Row, error) {
```
Comment (zachmu) [inline sql/plan/update.go]: This is your bug :D

File: sql/type.go [hunk 6]
```diff
@@ -386,6 +386,16 @@ func (t numberT) Compare(a interface{}, b interface{}) (int, error) {
 func (t numberT) String() string { return t.t.String() }
 
 func compareFloats(a interface{}, b interface{}) (int, error) {
+	aIsNull := a == nil
```
Comment (zachmu) [inline sql/type.go]: This looks correct but you need tests for this change

File: engine_test.go (hunk 2 above)
Comment (zachmu) [reply engine_test.go]: I stand corrected:

https://dev.mysql.com/doc/refman/8.0/en/update.html

I guess we do what MySQL does, even though it's wrong :(

Comment (Hydrocharged) [reply engine_test.go]: _Non-standard_ 😄

**Comment by Hydrocharged** [comment]
Alright, this is now ready for the final review pass. That ORDER BY was a great idea to add to the tests because I found a bug through it. Was a pain to debug though (as is anything in the pipeline).

File: sql/core.go [hunk 7]
```diff
@@ -217,6 +217,11 @@ type Replacer interface {
 	Inserter
 }
 
+type Updater interface {
+	// Update the given row. Provides both the old and new rows.
```
Comment (zachmu) [inline sql/core.go]: Probably should describe any special errors implementors are expected to return, if any.

File: sql/analyzer/resolve_orderby.go [hunk 8]
```diff
@@ -57,7 +59,15 @@ func resolveOrderBy(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error)
 		// below its child.
 		if len(colsFromChild) == 0 && len(missingCols) > 0 {
 			a.Log("pushing down sort, missing columns: %s", strings.Join(missingCols, ", "))
-			return pushSortDown(sort)
+			newNode, err := pushSortDown(sort)
+			if err != nil {
+				return newNode, err
+			}
+			_, ok := newNode.(*plan.ResolvedTable)
```
Comment (zachmu) [inline sql/analyzer/resolve_orderby.go]: What happens if you remove this check? Does the execution plan not catch the unresolved columns in some other way later on?

File: sql/type.go [hunk 9]
```diff
@@ -1173,3 +1209,16 @@ func convertArrayForJSON(t arrayT, v interface{}) (interface{}, error) {
 		return nil, ErrNotArray.New(v)
 	}
 }
+
+func hasNullInputs(a interface{}, b interface{}) (bool, int) {
```
Comment (zachmu) [inline sql/type.go]: Would call this compareNulls and document the return values and behavior with a method comment.

File: sql/type.go [hunk 10]
```diff
@@ -386,6 +386,10 @@ func (t numberT) Compare(a interface{}, b interface{}) (int, error) {
 func (t numberT) String() string { return t.t.String() }
 
 func compareFloats(a interface{}, b interface{}) (int, error) {
+	if hasNulls, res := hasNullInputs(a, b); hasNulls {
```
Comment (zachmu) [inline sql/type.go]: I would break this null sorting change into its own PR and feed it back to them separately, since it's independent of the new update feature.

File: sql/analyzer/resolve_orderby.go (hunk 8 above)
Comment (Hydrocharged) [reply sql/analyzer/resolve_orderby.go]: Correct, the ORDER BY doesn't fail or anything, it essentially just gets ignored. Comparing the behavior with MySQL, it's supposed to error out, so this is that check.

File: sql/core.go (hunk 7 above)
Comment (Hydrocharged) [reply sql/core.go]: No special errors are expected to be returned. `Delete` is the only one that returns a special error.

File: sql/type.go (hunk 9 above)
Comment (Hydrocharged) [reply sql/type.go]: Great idea!

File: sql/type.go (hunk 10 above)
Comment (Hydrocharged) [reply sql/type.go]: Moving this to its own PR would cause the related tests in this PR to fail. Is that okay? This change is fairly core to the correct behavior of `Update` working as one would expect.
